Implement qdigest type and associated functions

tdcmeehan opened the pull request:
Adds the following functions for creating and manipulating qdigests:
- `qdigest_agg`: aggregation function to create qdigest from real, bigint or double
  values
- `merge`: aggregation function to merge multiple qdigests
- `value_at_quantile`: scalar function that retrieves a given quantile from a
  qdigest
- `values_at_quantiles`: scalar function that retrieves a list of quantiles from a
  qdigest

Weights and accuracy specifications can be set to `qdigest_agg` in the same way they can be set on `approx_percentile`.  `qdigest`s support casting to and from `varbinary`.

Fixes #8695 

The serialization format still needs to be agreed on and documented, so this should not be merged until that is done.  airlift/airlift#678

tdcmeehan commented:
Note whereas this was allowed before:

`merge(null)`

It will no longer be allowed as null could refer to `HyperLogLog` or `qdigest` types.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MergeQuantileDigestFunction.java:
```diff
@@ -0,0 +1,150 @@
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE),
+                new ParameterMetadata(BLOCK_INPUT_CHANNEL, valueType),
+                new ParameterMetadata(BLOCK_INDEX));
```
We generally put each argument on its own line. Format as

```
        return ImmutableList.of(
                new ParameterMetadata(STATE),
                new ParameterMetadata(BLOCK_INPUT_CHANNEL, valueType),
                new ParameterMetadata(BLOCK_INDEX));

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/QuantileDigestAggregationFunction.java:
```diff
@@ -0,0 +1,210 @@
+        Type intermediateType = stateSerializer.getSerializedType();
+
+        checkArgument(inputFunctions.containsKey(valueType.getDisplayName()),
+                "Unsupported type %s supplied, known types: %s", valueType.getDisplayName(), inputFunctions.keySet());
```
Declare the 3 method handles are static final fields of this class, and use `switch ... case` to find the appropriate one here. Do not use a map.

This is how code like this have generally been written. There were careful benchmarks. However, I don't remember whether this style is critical to performance.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/QuantileDigestAggregationFunction.java:
```diff
@@ -0,0 +1,210 @@
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(QuantileDigestAggregationFunction.class, "combineState", QuantileDigestState.class, QuantileDigestState.class);
+    private static final MethodHandle OUTPUT_FUNCTION = methodHandle(QuantileDigestAggregationFunction.class, "evaluateFinal", QuantileDigestStateSerializer.class, QuantileDigestState.class, BlockBuilder.class);
+
+    public QuantileDigestAggregationFunction(TypeSignature... typeSignatures)
```
make this private

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/QuantileDigestAggregationFunction.java:
```diff
@@ -0,0 +1,210 @@
+        if (arity > 2) {
+            builder.add(DOUBLE);
+        }
+        return builder.build();
```
I would like this instead for simplicity

```
switch (arity) {
  case 1:
    return ImmutableList.of(valueType);
  case 2:
    // weight
    return ImmutableList.of(valueType, BIGINT);
  case 3:
    // weight and accuracy
    return ImmutableList.of(valueType, BIGINT, DOUBLE);
  default:
    throw ...
}

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/QuantileDigestAggregationFunction.java:
```diff
@@ -0,0 +1,210 @@
+        // accuracy unspecified
+        if (arity < 3) {
+            return insertArguments(inputFunction, 3, DEFAULT_ACCURACY);
+        }
```
Use switch instead (and add sanity check along the way).

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/QuantileDigestAggregationFunction.java:
```diff
@@ -0,0 +1,210 @@
+        input(state, floatToSortableInt(intBitsToFloat((int) value)), weight, accuracy);
+    }
+
+    public static void input(QuantileDigestState state, long value, long weight, double accuracy)
```
call this `inputBigint`

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/QuantileDigestFunctions.java:
```diff
@@ -0,0 +1,115 @@
+
+    private QuantileDigestFunctions() {}
+
+    @ScalarFunction("value_at")
```
I personally find that the `extract_quantile` in your original proposal is much better

tdcmeehan replied:
(As we discussed offline.)  How about `value_at_quantile`

haozhun commented on presto-docs/src/main/sphinx/functions/qdigest.rst:
```diff
@@ -0,0 +1,49 @@
+and it takes a parameter which must be one of ``bigint``, ``double``
+or ``real``.  A ``qdigest`` can be used to store approximate percentile
+information from a data set.  They may also be merged without losing
+precision.
```
I would explain here how one could materialize this to storage (by casting to/from varbinary)

haozhun commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -6958,6 +6964,274 @@ public void testApproxSetDouble()
         assertEquals(actual.getMaterializedRows(), expected.getMaterializedRows());
     }
 
+    @Test
```
I'll explain to you offline how I would write the tests for this PR.

tdcmeehan commented:
@haozhun thank you for the review.  Back to you.

dain commented:
Make sure to finish syntax review before merging.

wenleix commented:
Make sure to wrap commit message body at 72 characters before merging.

tdcmeehan commented:
Squashed and rebased (mindful of the git commit guidelines), and I'm triple confirming we're all happy on the syntax.  Presuming that's all OK this depends on a release of airlift to get the updated serialization format.

tdcmeehan commented:
Depends on #11898

tdcmeehan commented:
Reviewed syntax above one last time with @martint.

tdcmeehan commented:
Discussed with @ggreg , given this is an isolated change, I am merging now that airlift update is also included.

wenleix commented on presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestMergeQuantileDigestFunction.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    // The following tests are overridden because by default simple equality checks are done, which often won't work with
+    // qdigests due to the way they are serialized.  I am instead overridding these methods and using the QDIGEST_EQUALITY
```
nit: Use passive voice:
```
Instead, these methods are overridden and  QDIGEST_EQUALITY function is used to perform equality checks. 
```
